### PR TITLE
Developing the `openqasm3` to `QIR` visitor

### PR DESCRIPTION
**Pull request by TheGupta2012**
*issue tag(s), if exist*:

## Changes
- Developing the openqasm3 converter
- Using a tree-based approach to decompose the `openqasm3` AST into pyqir constructs.
- Progress - 
    - [x] Register Declarations
    - [x] QuantumMeasurements
    - [x] QuantumReset
    - [x] QuantumGates - `pyqir._native` support + `U[x]` gates
    - [x] QuantumBarriers 
    - [x] Gate Unfolding
    - [x] Custom Quantum Gates
    - [x] Simple Branching Statements (controlled on 1 bit)
    - [x] Expressions
        - [x] Literals

### CODE REVIEW

**Comment by TheGupta2012** [comment]
Hi @ryanhill1, I think the code is ready for the `v0.5` release. 
Once you review and merge in `main`, I will open up different PRs to tackle different aspects of the work that is left.

File: qbraid_qir/qasm3/visitor.py [hunk 1]
```diff
@@ -0,0 +1,828 @@
...
+            # we only support static integer sizes
+            if isinstance(size, IntegerLiteral):
+                value = size.value
+            else:
+                raise Qasm3ConversionError(f"Unsupported integer size {size} in {statement}")
+            var_name = statement.identifier.name
+
+            # how to add this in the QIR???
+        else:
+            raise Qasm3ConversionError(f"Unsupported classical type {decl_type} in {statement}")
```
Comment (ryanhill1) [inline qbraid_qir/qasm3/visitor.py:631]: What's going on in this function? If not implemented, we can just raise `NotImplementedError`. Either way, will want to clean this up a bit before we merge.

**Review by ryanhill1 — CHANGES_REQUESTED** [review]
Looks great! If you could just take a look and clean up that one function that I commented on. Other than that, this looks ready to merge.

File: qbraid_qir/qasm3/visitor.py (hunk 1 above)
Comment (TheGupta2012) [reply qbraid_qir/qasm3/visitor.py:631]: I was exploring how to define classical variables like `int`, `bool`, etc. but have not implemented it yet. Will just raise `NotImplemented` error as of now

File: qbraid_qir/qasm3/elements.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+class _Register(_ProgramElement):
+    def __init__(self, register: Tuple[str, Optional[int]], is_qubit: bool = True):
+        self._register = register
+        self._is_qubit = is_qubit
+
+    def accept(self, visitor):
+        visitor.visit_register(self._register, self._is_qubit)
+
+    def __str__(self) -> str:
+        return f"Register({self._register}, is_qubit = {self._is_qubit})"
```
[the remark this answers is not in the document]
Comment (TheGupta2012) [reply qbraid_qir/qasm3/elements.py:75]: Yea, I agree, using the `openqasm` constructs makes it more succinct.

**Review by ryanhill1 — APPROVED** [review]

**Pull request merged**